binary-oracle-pair/program/src/processor.rs

```diff
@@ -284,10 +331,183 @@ impl Processor {
+        let token_fail_mint_info = next_account_info(account_info_iter)?;
+        let user_token_destination_account_info = next_account_info(account_info_iter)?;
+        let token_program_id_info = next_account_info(account_info_iter)?;
+
+        if amount == 0 {
+            return Err(PoolError::InvalidAmount.into());
+        }
+
+        let user_pass_token_account = Account::unpack(&token_pass_user_account_info.data.borrow())?;
+        let user_fail_token_account = Account::unpack(&token_fail_user_account_info.data.borrow())?;
+        let user_token_destination_account =
+            Account::unpack(&user_token_destination_account_info.data.borrow())?;
+
+        if user_pass_token_account.owner != user_fail_token_account.owner
+            && user_pass_token_account.owner != user_token_destination_account.owner
+        {
+            return Err(PoolError::InvalidAccountsOwner.into());
+        }
+
+        let pool = Pool::unpack(&pool_account_info.data.borrow())?;
+
+        let authority_pub_key =
+            Self::authority_id(program_id, pool_account_info.key, pool.bump_seed)?;
+        if *authority_account_info.key != authority_pub_key {
+            return Err(PoolError::InvalidAuthorityAccount.into());
+        }
+
+        if pool.decision == None {
+            let possible_withdraw_amount = cmp::min(
+                user_pass_token_account.amount,
+                user_fail_token_account.amount,
+            );
+            if possible_withdraw_amount < amount {
+                return Err(PoolError::InsufficientFunds.into());
+            }
+
+            // Burn PASS tokens
+            Self::burn(
+                &pool_account_info.key,
+                pool.bump_seed,
+                token_program_id_info.clone(),
+                token_pass_mint_info.clone(),
+                token_pass_user_account_info.clone(),
+                authority_account_info.clone(),
+                possible_withdraw_amount,
+            )?;
+
+            // Burn FAIL tokens
+            Self::burn(
+                &pool_account_info.key,
+                pool.bump_seed,
+                token_program_id_info.clone(),
+                token_fail_mint_info.clone(),
+                token_fail_user_account_info.clone(),
+                authority_account_info.clone(),
+                possible_withdraw_amount,
+            )?;
+
+            // Transfer deposit tokens from pool deposit account to user destination account
+            Self::transfer(
+                &pool_account_info.key,
+                pool.bump_seed,
+                user_token_destination_account_info.clone(),
+                authority_account_info.clone(),
+                token_program_id_info.clone(),
+                pool_deposit_token_account_info.clone(),
+                possible_withdraw_amount,
+            )?;
+        } else if pool.decision == Some(false) {
+            if user_fail_token_account.amount < amount {
```

Comment:
No need for this check, should be done automatically by  the `burn` instruction.